### PR TITLE
Add additional result builder capabilities.

### PR DESCRIPTION
**Pull request by connor-ricks**
## Description

This PR extends the supported functionality of the `MiddlewareFixedTypeBuilder` result builder.

- Adds support for conditional `if-else` statement syntax in result builders
- Adds support for conditional `if` statement syntax in result builders
- Adds support for `optional` unwrapping statement syntax in result builders.
- Adds support for `for-loop` statement syntax in result builders
- Adds tests for validating the additional supported syntax.

The addition of these result builders also empowers `HummingbirdRouter` unlocking the ability for `RouteBuilder` and `RouterController` objects to make use of this conditional syntax as well.

> [!Warning]
It might be necessary to add tests to validate that `HummingbirdRouter` can also make use of these syntax additions. Although, given that we have tests that cover the ability to use controllers in the `MiddlewareFixedTypeBuilder` result builder, the support may be "inherently" tested.

### Conditional  `if-else`, `if` and `optional` statements

Adding support for conditional statements in result builders further enhances the ability to perform logical operations when composing your application's routes and middleware. 

#### `if-else`
Some applications may want to conditionally add different middleware based on various properties or configurations of their application. In the below example, we utilize different logging middleware if we are running in release mode.

```swift
router.addMiddleware {
    if isRelease {
        ReleaseLoggingMiddleware()
    } else {
        DebugLoggingMiddleware()
    }
    ...
}
```

#### `if`
Some applications may want to conditionally add middleware based on various properties or configurations of their application. In the below example, we attach a middleware that exposes the internal errors of our application to aid in client side debugging during development.

```swift
router.addMiddleware {
    if isDevelopment {
        DebugErrorDescriptionMiddleware()
    }
    ...
}
```

#### `optionals`
Sometimes our middleware builders may have an optional middleware that they would like to render. In the below example, we build our middleware in a function that takes an optional `logger`. We unwrap the middleware and add it to the router if it exists.

```swift
func buildMiddleware<M: RouterMiddleware>(logger: M?) async throws where M.Context == BasicRequestContext {
    self.router.addMiddleware {
        if let logger {
            logger
        }
        ...
    }
}
```

#### `for-loop`
Sometimes we may build middleware based on a provided set of data. In the below example, we add a series of tracking middleware based on the provided collection of `Tracker` objects.

```swift
enum Tracker: String {
    case sentry
    case datadog
    case firebase
}

func add(trackers: [Tracker]) {
    self.router.addMiddleware {
        for tracker in trackers {
            switch tracker {
            case .datadog:
                DatadogMiddleware()
            case .firebase:
                FirebaseMiddleware()
            case .sentry:
                SentryMiddleware()
            }
        }
    }
}
```

### CODE REVIEW

File: Sources/Hummingbird/Middleware/MiddlewareModule/MiddlewareFixedTypeBuilder.swift [hunk 1]
```diff
@@ -12,34 +12,6 @@
 //
 //===----------------------------------------------------------------------===//
 
-// MARK: - Middleware2
-
-public struct _Middleware2<M0: MiddlewareProtocol, M1: MiddlewareProtocol>: MiddlewareProtocol where M0.Input == M1.Input, M0.Context == M1.Context, M0.Output == M1.Output {
```
Comment (connor-ricks) [inline Sources/Hummingbird/Middleware/MiddlewareModule/MiddlewareFixedTypeBuilder.swift:17]: Moved to `_Middleware2.swift`

File: Sources/Hummingbird/Middleware/MiddlewareModule/MiddlewareFixedTypeBuilder.swift [hunk 2]
```diff
@@ -64,4 +36,28 @@ public enum MiddlewareFixedTypeBuilder<Input, Output, Context> {
...
+            .first(component)
+        } else {
+            .second(_EmptyMiddleware())
+        }
+    }
+
+    public static func buildEither<M0: MiddlewareProtocol, M1: MiddlewareProtocol>(
+        first content: M0
+    ) -> _ConditionalMiddleware<M0, M1> {
+        .first(content)
```
Comment (adam-fowler) [inline Sources/Hummingbird/Middleware/MiddlewareModule/MiddlewareFixedTypeBuilder.swift:51]: Why do we need the `ConditionalMiddleware` type here? Can't you just return `content`? eg
```swift
public static func buildEither<M0: MiddlewareProtocol>(
    first: M0
) -> M0 {
    first
}

public static func buildEither<M1: MiddlewareProtocol>(
    second: M1
) -> M1 {
    second
}
```

File: Sources/Hummingbird/Middleware/MiddlewareModule/MiddlewareFixedTypeBuilder.swift [hunk 3]
```diff
@@ -64,4 +36,28 @@ public enum MiddlewareFixedTypeBuilder<Input, Output, Context> {
     ) -> _Middleware2<M0, M1> where M0.Input == M1.Input, M0.Output == M1.Output, M0.Context == M1.Context {
         _Middleware2(m0, m1)
     }
+
+    public static func buildOptional<M0: MiddlewareProtocol>(_ component: M0?) -> _ConditionalMiddleware<M0, _EmptyMiddleware<M0>> {
```
Comment (adam-fowler) [inline Sources/Hummingbird/Middleware/MiddlewareModule/MiddlewareFixedTypeBuilder.swift:40]: If `_ConditionalMiddleware` isn't used by `buildEither` then this is the only place it is used. Can we instead just write an `OptionalMiddleware` and get rid of `_ConditionMiddleware` and `_EmptyMiddleware`.

**Review by adam-fowler — COMMENTED** [review]
Some initial comments

File: Sources/Hummingbird/Middleware/MiddlewareModule/MiddlewareFixedTypeBuilder.swift (hunk 2 above)
Comment (connor-ricks) [reply Sources/Hummingbird/Middleware/MiddlewareModule/MiddlewareFixedTypeBuilder.swift]: 🤔 I think in some of my personal result builders this was not possible, but yea in this case it is.

File: Sources/Hummingbird/Middleware/MiddlewareModule/MiddlewareFixedTypeBuilder.swift (hunk 3 above)
Comment (connor-ricks) [reply Sources/Hummingbird/Middleware/MiddlewareModule/MiddlewareFixedTypeBuilder.swift]: Good feedback! 👍 I've pushed changes to address the recommendations.

File: Sources/Hummingbird/Middleware/MiddlewareModule/_Middleware2.swift [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+///
+/// You won't typically construct this middleware directly, but instead will use result builder syntax.
+///
+/// ```swift
+/// router.addMiddleware {
+///    MiddlewareOne()
+///    MiddlewareTwo()
+/// }
+/// ```
+public struct _Middleware2<M0: MiddlewareProtocol, M1: MiddlewareProtocol>: MiddlewareProtocol where M0.Input == M1.Input, M0.Context == M1.Context, M0.Output == M1.Output {
```
Comment (adam-fowler) [inline Sources/Hummingbird/Middleware/MiddlewareModule/_Middleware2.swift:27]: Can you add `@_documentation(visibility: internal)`. This isn't a symbol we need to publish documentation for.

File: Sources/Hummingbird/Middleware/MiddlewareModule/_OptionalMiddleware.swift [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+///
+/// ```swift
+/// router.addMiddleware {
+///   if let middleware {
+///     middleware
+///   }
+///   ...
+/// }
+/// ```
+public struct _OptionalMiddleware<M0: MiddlewareProtocol>: MiddlewareProtocol {
```
Comment (adam-fowler) [inline Sources/Hummingbird/Middleware/MiddlewareModule/_OptionalMiddleware.swift:30]: Can you add `@_documentation(visibility: internal)`. This isn't a symbol we need to publish documentation for.

File: Sources/Hummingbird/Middleware/MiddlewareModule/_SpreadMiddleware.swift [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+/// statements in a result builder to automatically build spread middleware:
+///
+/// ```swift
+/// router.addMiddleware {
+///   for logger in loggers {
+///     LoggingMiddleware(logger: logger)
+///   }
+/// }
+/// ```
+public struct _SpreadMiddleware<M0: MiddlewareProtocol>: MiddlewareProtocol {
```
Comment (adam-fowler) [inline Sources/Hummingbird/Middleware/MiddlewareModule/_SpreadMiddleware.swift:29]: Can you add `@_documentation(visibility: internal)`. This isn't a symbol we need to publish documentation for.

**Review by adam-fowler — CHANGES_REQUESTED** [review]
Some minor documentation changes to hide the different middleware from documentation, as I don't think they need to be there. Otherwise looks good

**Comment by connor-ricks** [comment]
Added. 👍  Thanks for the review!

**Review by adam-fowler — APPROVED** [review]
Looks good
We can ignore the breaking change, which seems to apply to adding the `@documentation` tag to a symbol

**Pull request merged**